QueryBuilder/Common/Statements/WhereStatement.cs

```diff
@@ -0,0 +1,429 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Common.Statements
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Linq;
+    using System.Reflection;
+    using System.Runtime.Serialization;
+    using Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Common.Clauses;
+    using Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Common.Helpers;
+
+    /// <summary>
+    /// An abstract class that provides the base layer of methods that are common between
+    /// WHERE statement classes that are applicable to either Twins or Relationships.
+    /// </summary>
+    /// <typeparam name="TWhereStatement">The type of WHERE statement supported.</typeparam>
+    public abstract class WhereBaseStatement<TWhereStatement> where TWhereStatement : WhereBaseStatement<TWhereStatement>
+    {
+        /// <summary>
+        /// The alias for the current WHERE statement.
+        /// </summary>
+        protected readonly string Alias;
+        internal readonly WhereClause WhereClause;
+        internal readonly JoinOptions JoinOptions;
+
+        internal WhereBaseStatement(JoinOptions joinOptions, WhereClause clause, string alias) : this(clause, alias)
+        {
+            JoinOptions = joinOptions;
+        }
+
+        internal WhereBaseStatement(WhereClause whereClause, string alias)
+        {
+            WhereClause = whereClause;
+            Alias = alias;
+        }
+
+        /// <summary>
+        /// Wraps a given set of WHERE conditions in parenthesis to establish a precedence.
+        /// </summary>
+        /// <param name="nested">The functional logic of the WHERE statement containing one or more WHERE conditions.</param>
+        /// <returns>A conjunction class that supports appending more conditions to the WHERE statements via OR or AND terms.</returns>
+        public WhereCombineStatement<TWhereStatement> Precedence(Func<TWhereStatement, WhereCombineStatement<TWhereStatement>> nested)
+        {
+            var statement = ActivatorHelper.CreateInstance<TWhereStatement>(JoinOptions, new WhereClause(), Alias);
+            var n = nested.Invoke(statement);
+            WhereClause.AddCondition($"({n.WhereClause.GetConditionsText()})");
```

Comment:
I don't disagree, but how we do this without changing how the existing WhereClause works for the typed builder is challenging, so this approach seemed to be the easiest to avoid adversely affecting that flow.